dial.go

```diff
@@ -170,6 +170,17 @@ func (d *Dialer) Dial(ctx context.Context, raddr ma.Multiaddr, remote peer.ID) (
 		return nil, errOut
 	}
 
+	// if the connection is not to whom we thought it would be...
+	connRemote := connOut.RemotePeer()
+	if connRemote != remote {
+		connOut.Close()
+		_, err := connOut.Read(nil) // should return any potential errors (ex: from secio)
```

Comment:
This really shouldn't be necessary. At this point, the secio handshake has completed successfully (we just might not be connected to the right peer). 

We should probably log the error from `Close()`, if anything.